src/libsystemd/sd-device/device-enumerator.c

```diff
@@ -313,6 +405,11 @@ static int enumerator_sort_devices(sd_device_enumerator *enumerator) {
 
         enumerator->sorted = true;
         return 0;
+
+failed:
+        device_unref_many(devices, n);
+        free(devices);
+        return r;
```

Comment:
hmm, so what about an algorithm like this: we first generate a 1:1 array with all devices. And then we go through that list once for each priorized subsystem: if the device we look at is from the priorized subsystem we move it the front, swapping the position with the first device that we haven't processed so far.

```c
size_t p = 0;
STRV_FOREACH(prioritized_subsystem, enumerator->prioritized_subsystems) {
        for (size_t i = 0; i < n; i++) {
                if (!device_matches_subsystem(devices[i], *prioritized_subsystem))
                        continue;
                SWAP_TO(devices[i], devices[p]);
                p++;
        }

        qsort(devices + n_sorted, p - n_sorted);
}
```
(pseudo-code)

with that logic we wouldn#t have to remove/readd anything to the hashmap, we'd just reorder the array in neat partitions, and then sort each partition.

wouldn't thta work?

Reply:
I think your proposal is inefficient, as `device_matches_subsystem()` must search its all child devices and check their subsystems match `*prioritized_subsystem`. Consider the case that a sd-device object which refers to a PCI bus which has a block device in the tree...